include/swift/Frontend/ParseableInterfaceSupport.h

```diff
@@ -55,6 +56,65 @@ bool emitParseableInterface(raw_ostream &out,
 std::string
 getModuleCachePathFromClang(const clang::CompilerInstance &Instance);
 
+
+struct ForwardingModule {
+  std::string underlyingModulePath;
+  struct Dependency {
+    std::string path;
+    uint64_t size;
+    uint64_t lastModificationTime;
+  };
+  std::vector<Dependency> dependencies;
+  ForwardingModule() = default;
+  ForwardingModule(StringRef underlyingModulePath)
+    : underlyingModulePath(underlyingModulePath) {}
+
+  void addDependency(StringRef path, uint64_t size, uint64_t modTime) {
+    dependencies.push_back({path.str(), size, modTime});
+  }
+};
+
+class ModuleDiscovery {
```

Comment:
These need to be documented. Eventually with a big diagram of how the cache works, and how the fallback mechanism works.